### PR TITLE
[Enhancement] make hdfs cache pool configurable

### LINKED ISSUE #29019: BE crashes when querying hive table on HDFS (on hdfsGetHedgedReadMetrics)
### Steps to reproduce the behavior (Required)

I've tested queries on multiple hive tables, and BE crashed with following stack

```
*** SIGSEGV (@0x180003fd21) received by PID 183457 (TID 0x7f9a31663700) from PID 261409; stack trace: ***
    @          0x5f6cd02 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f9af240dcab os::Linux::chained_handler()
    @     0x7f9af241259c JVM_handle_linux_signal
    @     0x7f9af24058f8 signalHandler()
    @     0x7f9af18f05e0 (unknown)
    @     0x7f9af21d3c92 Fingerprinter::Fingerprinter()
    @     0x7f9af21ca569 jni_invoke_nonstatic()
    @     0x7f9af21cb78a jni_CallObjectMethodV
    @          0x58f6a23 invokeMethodOnJclass
    @          0x58f6d6f invokeMethod
    @          0x58f8b66 hdfsGetHedgedReadMetrics
    @          0x407879d _ZZN9starrocks15HdfsInputStream22get_numeric_statisticsEvENKUlvE_clEv
    @          0x4078952 _ZNSt17_Function_handlerIFN9starrocks6StatusEvEZNS0_15HdfsInputStream22get_numeric_statisticsEvEUlvE_E9_M_invokeERKSt9_Any_data
    @          0x4ac633e starrocks::call_hdfs_scan_function_in_pthread()
    @          0x4079bb4 starrocks::HdfsInputStream::get_numeric_statistics()
    @          0x2df33f7 starrocks::io::SeekableInputStreamWrapper::get_numeric_statistics()
    @          0x4124ff7 starrocks::io::SharedBufferedInputStream::get_numeric_statistics()
    @          0x2df33f7 starrocks::io::SeekableInputStreamWrapper::get_numeric_statistics()
    @          0x2df33f7 starrocks::io::SeekableInputStreamWrapper::get_numeric_statistics()
    @          0x5248672 starrocks::HdfsScanner::update_hdfs_counter()
    @          0x5248b08 starrocks::HdfsScanner::update_counter()
    @          0x524976c starrocks::HdfsScanner::close()
    @          0x51ae89e starrocks::connector::HiveDataSource::close()
    @          0x28ea8a0 starrocks::pipeline::ScanOperator::_close_chunk_source_unlocked()
    @          0x28e97e8 starrocks::pipeline::ScanOperator::_finish_chunk_source_task()
    @          0x28edb35 _ZZN9starrocks8pipeline12ScanOperator18_trigger_next_scanEPNS_12RuntimeStateEiENKUlvE_clEv
    @          0x28f63b1 starrocks::workgroup::ScanExecutor::worker_thread()
    @          0x4b96ca2 starrocks::ThreadPool::dispatch_thread()
    @          0x4b9179a starrocks::Thread::supervise_thread()
    @     0x7f9af18e8e25 start_thread
    @     0x7f9af0f0834d __clone
    @                0x0 (unknown)
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by dirtysalt**
Fixes https://github.com/StarRocks/starrocks/issues/29019

This PR can reduce the chance of that issue.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Review by DorianZheng — APPROVED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

**Review by ZiheLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by stephen-shelby — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/fs/hdfs/hdfs_fs_cache.cpp [hunk 1]
```diff
@@ -106,26 +106,26 @@ Status HdfsFsCache::get_connection(const std::string& namenode, std::shared_ptr<
...
-        uint32_t idx = _rand.Uniform(_max_cache_clients);
-        _cache_key[idx] = cache_key;
+    if (UNLIKELY(_cur_client_idx >= max_cache_clients)) {
+        uint32_t idx = _rand.Uniform(max_cache_clients);
+        _cache_keys[idx] = cache_key;
         _cache_clients[idx] = hdfs_client;
     } else {
-        _cache_key[_cur_client_idx] = cache_key;
-        _cache_clients[_cur_client_idx] = hdfs_client;
-        _cur_client_idx++;
```
Comment (imay) [inline be/src/fs/hdfs/hdfs_fs_cache.cpp:128]: Why do you remove `_cur_client_idx++`? If you remove it, when _cur_client_idx will be changed?

Comment (dirtysalt) [reply be/src/fs/hdfs/hdfs_fs_cache.cpp:128]: because we use vector(instead of array) to store cache key and client.

**Review by satanson — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Comment by dirtysalt** [comment]
Dear, I overlook @imay's proposal. And actually there is a bug. fixed in this pr https://github.com/StarRocks/starrocks/pull/29037

**Comment by dirtysalt** [comment]
Don't backport this PR directly.

**Pull request merged**